openedx_learning/apps/authoring/publishing/api.py

```diff
@@ -216,10 +216,20 @@ def get_all_drafts(learning_package_id: int, /) -> QuerySet[Draft]:
     )
 
 
-def get_entities_with_unpublished_changes(learning_package_id: int, /) -> QuerySet[PublishableEntity]:
-    return PublishableEntity.objects \
-                            .filter(learning_package_id=learning_package_id) \
-                            .exclude(draft__version=F('published__version'))
+def get_entities_with_unpublished_changes(
+    learning_package_id: int,
+    /,
+    include_deleted_drafts: bool = False
+) -> QuerySet[PublishableEntity]:
+    """
+    Fetch entities that have unpublished changes.
+
+    By default, this excludes soft-deleted drafts but can be included using include_deleted_drafts option.
+    """
+    query_filters = {"learning_package_id": learning_package_id}
+    if not include_deleted_drafts:
+        query_filters['draft__version__isnull'] = False
+    return PublishableEntity.objects.filter(**query_filters).exclude(draft__version=F('published__version'))
```

Comment:
This will erroneously return soft-deleted entries that have already been published, because of the NULL comparisons. I think that if we're making this flag, it needs to be something like:

```python
entities_qs = (
    PublishableEntity.objects
    .filter(learning_package_id=learning_package_id)
    .exclude(draft__version=F('published__version'))
)

if include_deleted_drafts:
    # This means that we should also return PublishableEntities where the draft
    # has been soft-deleted, but that deletion has not been published yet. Just
    # excluding records where the Draft and Published versions don't match won't
    # be enough here, because that will return soft-deletes that have already
    # been published (since NULL != NULL in SQL).
    # 
    # So we explcitly exclude already-published deletes:
    return entities_qs.exclude(
        Q(draft__version__isnull=True) & Q(published__version__isnull=True)
    )

# Simple case: filter out the entities that have been soft-deleted.
return entities_qs.filter(draft__version__isnull=False)
```

(I had to deal with this in https://github.com/openedx/openedx-learning/pull/211, which I'd love to get a review of if you have a moment.)

Reply:
@ormsbee Makes sense. Updated and included a test inspired by your PR: #211